pystrometry/utils/du437_tools.py

```diff
@@ -697,3 +700,158 @@ def show_best_solution(file, out_dir):
+
+    p = (A ** 2 + B ** 2 + G ** 2 + F ** 2) / 2.
+    q = A * G - B * F
+
+    a_mas = unp.sqrt(p + unp.sqrt(p ** 2 - q ** 2))
+    i_rad = unp.arccos(q / (a_mas ** 2.))
+    omega_rad = (unp.arctan2(B - F, A + G) + unp.arctan2(-B - F, A - G)) / 2.
+    OMEGA_rad = (unp.arctan2(B - F, A + G) - unp.arctan2(-B - F, A - G)) / 2.
+
+    # Convert radians to degrees
+    i_deg = i_rad * 180 / np.pi
+    omega_deg = omega_rad * 180 / np.pi
+    OMEGA_deg = OMEGA_rad * 180 / np.pi
+
+    # Extract nominal values and standard deviations
+    geometric_parameters = np.array([unp.nominal_values(a_mas), 
+                                     unp.nominal_values(omega_deg), 
+                                     unp.nominal_values(OMEGA_deg), 
+                                     unp.nominal_values(i_deg)])
+
+    geometric_parameters_errors = np.array([unp.std_devs(a_mas), 
+                                            unp.std_devs(omega_deg), 
+                                            unp.std_devs(OMEGA_deg), 
+                                            unp.std_devs(i_deg)])
+
+    return geometric_parameters, geometric_parameters_errors
+
+
+def correlation_matrix(input_table, num_pars=None):
+    """ 
+    This function reads the corr_vec from the nss_two_body_orbit table
+    and converts it to a numpy array with the full correlation matrix. 
+    By default it works for the 12 parameter 'Orbital' solution but it can
+    easily be adjusted to other solutions by providing the number of parameters
+    for the desired solution in the num_pars argument.
+    
+    Parameters
+    ----------
+    input_table : pandas Series or dict
+        A single row from the nss_two_body_orbit table for the desired target.
+    num_pars : int
+        Number of parameters for the corresponding solution of the desired target.
+
+    Returns
+    -------
+    corr_mat : ndarray
+        Correlation matrix for the specified parameters.
+    """
+
+    corr_vec = input_table['corr_vec']
+    if num_pars is None:
+        # Use the 'Orbital' solution by default which has 12 free parameters
+        num_pars = 12
+
+    corr_mat = np.ones([num_pars, num_pars], dtype=float)
+
+    # Fill in lower triangle with corr_vec
+    corr_mat[np.tril_indices(num_pars, k=-1)] = corr_vec
+    # Fill in upper triangle with mirror from lower triangle
+    upper_triangle = np.triu_indices(num_pars, k=1)
+    corr_mat[upper_triangle] = corr_mat.T[upper_triangle]
+
+    return corr_mat
+
+
+def covariance_matrix(input_table, pars=None):
+    """ 
+    Creates the covariance matrix from the corr_vec from the nss_two_body_orbit 
+    table. By default it works for the 12  'Orbital' parameter solution but it can
+    easily be adjusted to other solutions by providing the relevant parameters
+    (in the same order that they are listed in the Gaia documentation). 
+    
+    Parameters
+    ----------
+    input_table : pandas Series or dict
```

Comment:
Should this not be a pandas DataFrame instead?

Reply:
The input would only be one row of a DataFrame corresponding to the desired target, that's why it would be a pandas Series.

Reply:
OK, got it. Thanks